065-topic-replication-factor-change.md

```diff
@@ -0,0 +1,298 @@
+  Cruise Control's task states: None.
+  ```sh
+  status:
+    conditions:
+    - lastTransitionTime: "2024-01-18T16:13:50.490918232Z"
+      status: "True"
+      type: Ready
+    observedGeneration: 2
+    replicasChange:
+      state: pending
+      targetReplicas: 2
+    topicName: my-topic
+  ```
+
+- **Ongoing**: In Cruise Control's task queue, but execution not started, or not completed. 
+  Cruise Control's task states: Active, InExecution.
+  ```sh
+  status:
+    conditions:
+    - lastTransitionTime: "2024-01-18T16:13:53.490918232Z"
+      status: "True"
+      type: Ready
+    observedGeneration: 3
+    replicasChange:
+      sessionId: 1aa418ca-53ed-4b93-b0a4-58413c4fc0cb
+      state: ongoing
+      targetReplicas: 2
+    topicName: my-topic
+  ```
+
+- **Completed**: Cruise Control's task execution completed (target replication factor reconciled).
+  Cruise Control's task states: Completed.
+  ```sh
+  status:
+    conditions:
+    - lastTransitionTime: "2024-01-18T16:13:58.490918232Z"
+      status: "True"
+      type: Ready
+    observedGeneration: 4
+    topicName: my-topic
+  ```
+
+- **Failed**: Cruise Control's request failed, the change remains pending, and it is periodically retried.
+  Cruise Control's task states: CompletedWithError.
+  ```sh
+  status:
+    conditions:
+    - lastTransitionTime: "2024-01-18T16:13:58.490918232Z"
+      status: "True"
+      type: Ready
+    observedGeneration: 4
+    replicasChange:
+      message: Change request failed, Cluster model not ready
+      state: pending
+      targetReplicas: 2
+    topicName: my-topic
+  ```
+
+### Reconciliation logic
+
+At runtime, the Topic Operator watches and periodically reconciles the configuration of all managed and unpaused KafkaTopic resources.
+Among the other configurations, the reconciliation logic detects `.spec.replicas` changes in KafkaTopic resources by comparing with the topic replication factor value in Kafka.
+
+A ReplicasChangeClient will be used to isolate the logic required to call the Cruise Control REST API from the rest of the reconciliation logic.
+The ReplicasChangeClient will support sending multiple replicas changes with a single request, and checking multiple tasks with a single request.
+
+On every reconciliation loop, when Cruise Control integration will be enabled and replicas changes are detected, the BatchingTopicController will use the ReplicasChangeClient for the following operations:
+
+1. Request pending replicas changes (uses `topic_configuration` endpoint, and returns topics with pending and ongoing changes).
+2. Check ongoing replicas changes (uses `user_tasks` endpoint, and returns topics with completed changes).
+
+The result of these operations will be used to update the KafkaTopic status.
+
+### Error handling
+
+In case of error, the replicas change will remain pending or ongoing (no status update), and will be indefinitely retried by the periodic reconciliation.
```

Comment:
I didn't understand this part: You described a `Failed` state, but you don't seem to use it here.

Reply:
I see why it is confusing. The failed state is reported as an error message. The change remains pending so that it can be retried periodically. Proposal updated.